walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,179 @@
+    if (!freeTime.isEmpty() && !optionalAttendeeEvents.isEmpty()) {
+      freeTime =
+          freeSlotsWithOptionalAttendees(optionalAttendeeEvents, freeTime, optionalAttendees);
+    }
+
+    return freeTime;
+  }
+
+  // Loop through all the events given and modify freeTime List
+  private void freeSlotsWithMandatoryAttendees(
+      LinkedList<TimeRange> freeTime,
+      Collection<Event> events,
+      LinkedList<Event> optionalAttendeeEvents,
+      Collection<String> optionalAttendees,
+      Collection<String> mandatoryAttendees) {
+
+    for (Event event : events) {
+      /* Check if the event has any optional attendees and
+        save it in the optionalAtttendeeEvents list which we will deal with
+        later in the freeSlotsWithOptionalAttendees method */
+      if (!Collections.disjoint(optionalAttendees, event.getAttendees()))
+        optionalAttendeeEvents.add(event);
+      if(!Collections.disjoint(mandatoryAttendees, event.getAttendees())) 
+        findConflictAndResolve(freeTime, event);
+    }
+
+  }
+
+  // Try checking for any time slots we can get for the meeting with optional attendees
+  private LinkedList<TimeRange> freeSlotsWithOptionalAttendees(
+      LinkedList<Event> optionalAttendeeEvents,
+      LinkedList<TimeRange> freeTime,
+      Collection<String> optionalAttendees) {
+    // Create a new OptionalFreeTime list so that we still have our time slots with mandatory
+    // attendees
+    LinkedList<TimeRange> optionalFreeTime = new LinkedList<>(freeTime);
+    /* Loop through all the events our optional attendees are attending and
+       find free time slots we can have for our meeting */
+    for (Event event : optionalAttendeeEvents) {
+      findConflictAndResolve(optionalFreeTime, event);
+    }
+    /* If we can't find any free time slots with optional attendees
+      then return the original freeTime slots */
+    if (optionalFreeTime.isEmpty()) return freeTime;
+
+    return optionalFreeTime;
+  }
+
+  // This method finds the overlapping free time slot and event; altering the freeTime so that it
+  // doesn't overlap anymore.
+  private void findConflictAndResolve(
+      LinkedList<TimeRange> time, Event event) {
+      // Go through all the free time slots and check from which one is the event overlapping
+      ListIterator<TimeRange> iterator = time.listIterator();
+      while (iterator.hasNext()) {
+        TimeRange freeSlot = iterator.next();
+        // When th eoverlapping free slot is found send our free time and event slot to get altered.
+        if (freeSlot.overlaps(event.getWhen())) {
+          updateFreeTime(freeSlot, event.getWhen(), time, iterator);
+          // After updating free time is done, check if the event ends in this free slots
+          if (event.getWhen().end() <= freeSlot.end()) break;
+        }
+    }
+  }
+
+  /* This method free time, free time slot and current event with our free time slot and the iterator
```

Comment:
Use "/** */" for Javadoc comments.

Reply:
Um.. okay so the comments before a method are Javadoc comments and are supposed to with written using "/** */" but the comments inside the method are better if they start with "//" instead of "/* */"......did I understand it right?

Reply:
be*

Reply:
Um.. okay so the comments before a method are Javadoc comments and are supposed to be written using "/** */" but the comments inside the method are better if they start with "//" instead of "/* */"......did I understand it right?

Reply:
Um.. okay so the comments before a method are Javadoc comments and are supposed to be written using "/** * /" but the comments inside the method are better if they start with "//" instead of "/ * */"......did I understand it right?